### PR TITLE
Introduce New Service Status Codes.

### PR DESCRIPTION
**Pull request by kaiwalyajoshi**
This PR enhances the existing `<service>/v1/health` endpoint with additional status codes that indicate which stage the service is in its lifecycle.

JIRA Ticket: [DCOS-49197](https://jira.mesosphere.com/browse/DCOS-49197)

### CODE REVIEW

**Comment by kaiwalyajoshi** [comment]
@kensipe absolutely, this won't ship without tests. Wanted to get out an initial version we can play with.

**Comment by kaiwalyajoshi** [comment]
@takirala @kvish this ready for review.

This endpoint is at `<service>/v1/status/check?verbose=true`, it pretty much supercedes all the functionality in `<service>/v1/health`.

I think its a valid question now if we want to just move all of the new status check functionality into the existing health endpoint.

File: sdk/scheduler/src/main/java/com/mesosphere/sdk/debug/ServiceStatusTracker.java [hunk 1]
```diff
@@ -0,0 +1,571 @@
...
+
+    return new ServiceStatusEvaluationStage(statusCode, reason);
+  }
+
+  private ServiceStatusEvaluationStage isDeploying() {
+
+    String reason;
+    Optional<ServiceStatusCode> statusCode;
+
+    //Get the deployment plan.
```
Comment (takirala) [inline sdk/scheduler/src/main/java/com/mesosphere/sdk/debug/ServiceStatusTracker.java]: Adding a space after all `//` would make me super happy :)

File: sdk/scheduler/src/main/java/com/mesosphere/sdk/debug/ServiceStatusTracker.java [hunk 2]
```diff
@@ -0,0 +1,571 @@
...
+    ServiceStatusCode(int statusCode) {
+      this.statusCode = statusCode;
+    }
+  }
+
+  private static final String RESULT_CODE_KEY = "value";
+
+  private static final String BACKUP_PLAN_REGEXP = ".*back.*";
+
+  private static final String RESTORE_PLAN_REGEXP = ".*restor.*";
```
[the remark this answers is not in the document]
Comment (kaiwalyajoshi) [reply sdk/scheduler/src/main/java/com/mesosphere/sdk/debug/ServiceStatusTracker.java]: yup, it cases like "restoring", "restoration" etc...

File: sdk/scheduler/src/main/java/com/mesosphere/sdk/debug/ServiceStatusTracker.java [hunk 3]
```diff
@@ -0,0 +1,571 @@
...
+      statusCodeReasons.put(isErrorCreating.getStatusReason());
+      statusCodeReasons.put(initializing.getStatusReason());
+      statusCodeReasons.put(deploymentComplete.getStatusReason());
+      statusCodeReasons.put(isDeploying.getStatusReason());
+      statusCodeReasons.put(isDegraded.getStatusReason());
+      statusCodeReasons.put(isRecovering.getStatusReason());
+      statusCodeReasons.put(isBackingUp.getStatusReason());
+      statusCodeReasons.put(isRestoring.getStatusReason());
+      statusCodeReasons.put(isUpgradeRollbackDowngrade.getStatusReason());
+      response.put("reasons:", statusCodeReasons);
```
[the remark this answers is not in the document]
Comment (kaiwalyajoshi) [reply sdk/scheduler/src/main/java/com/mesosphere/sdk/debug/ServiceStatusTracker.java]: Nope, totally a typo thanks!

**Comment by kaiwalyajoshi** [comment]
@kensipe @takirala @kvish 

I've added all the functionality of `<service>/v1/status/check?verbose=true` into the existing `<service>/v1/health?verbose=true` endpoint.

The PR comments above have been address alongside. Please have a look.

File: sdk/scheduler/src/main/java/com/mesosphere/sdk/framework/FrameworkRunner.java [hunk 4]
```diff
@@ -210,14 +212,18 @@ public void registerAndRunFramework(Persister persister, MesosEventClient mesosE
    */
   private void runSkeletonScheduler() {
     PlanManager uninstallPlanManager = DefaultPlanManager.createProceeding(EMPTY_DEPLOY_PLAN);
+
+    PlanCoordinator coordinator = new DefaultPlanCoordinator(Optional.empty(),
+        Arrays.asList(uninstallPlanManager));
```
Comment (takirala) [inline sdk/scheduler/src/main/java/com/mesosphere/sdk/framework/FrameworkRunner.java]:
```suggestion
        Collections.singletonList(uninstallPlanManager));
```

File: sdk/scheduler/src/main/java/com/mesosphere/sdk/http/endpoints/HealthResource.java [hunk 5]
```diff
@@ -1,79 +1,515 @@
...
 
-  /**
-   * Creates a new instance which determines health based on the deploy and recovery plans in the provided plan
-   * coordinator.
-   */
-  public HealthResource(PlanCoordinator planCoordinator, SchedulerConfig schedulerConfig) {
-    this(getDeploymentAndRecoveryManagers(planCoordinator), schedulerConfig);
+    private final int statusCode;
+
+    ServiceStatusCode(int statusCode) {
```
Comment (takirala) [inline sdk/scheduler/src/main/java/com/mesosphere/sdk/http/endpoints/HealthResource.java]: wdyt of adding the priority value from javadoc above as another field to this enum? Makes it easy to understand the priority mapping. Also lets us not type `1`, `2` etc in the `String.format(..` statements below.

File: sdk/scheduler/src/main/java/com/mesosphere/sdk/http/endpoints/HealthResource.java [hunk 6]
```diff
@@ -1,79 +1,515 @@
...
+    // Service unavailable if nothing else.
+    if (!serviceStatusCode.isPresent()) {
+      serviceStatusCode = Optional.of(ServiceStatusCode.SERVICE_UNAVAILABLE);
+    }
+
+    response.put(RESULT_CODE_KEY, serviceStatusCode.get().statusCode);
+
+    // Build the final response.
+    final Response returnResponse = ResponseUtils.jsonResponseBean(response, serviceStatusCode.get().statusCode);
+    return new ServiceStatusResult(serviceStatusCode, returnResponse);
```
Comment (takirala) [inline sdk/scheduler/src/main/java/com/mesosphere/sdk/http/endpoints/HealthResource.java]:
```suggestion
    return new ServiceStatusResult(
    	serviceStatusCode,
    	ResponseUtils.jsonResponseBean(response, serviceStatusCode.get().statusCode)
    );
```

File: sdk/scheduler/src/main/java/com/mesosphere/sdk/http/endpoints/HealthResource.java [hunk 7]
```diff
@@ -1,79 +1,515 @@
...
  */
 @Path("/v1/health")
 public class HealthResource {
+
   /**
-   * We store the PlanManagers, not the underlying Plans, because PlanManagers can change their plans at any time.
+   * ServiceStatusCode encodes the service status code.
    */
-  private final Collection<PlanManager> planManagers;
+  public enum ServiceStatusCode {
```
Comment (takirala) [inline sdk/scheduler/src/main/java/com/mesosphere/sdk/http/endpoints/HealthResource.java]:
```suggestion
 enum ServiceStatusCode {
```
is there a downside of making this package-private ?

File: sdk/scheduler/src/main/java/com/mesosphere/sdk/http/endpoints/HealthResource.java [hunk 8]
```diff
@@ -1,79 +1,582 @@
...
+            .getPlan();
+
+    return evaluatePendingOrStartingStatusCode(deploymentPlan,
+        ServiceStatusCode.DEPLOYING_PENDING,
+        ServiceStatusCode.DEPLOYING_STARTING,
+        ServiceStatusCode.DEPLOYING_PENDING.priority);
+  }
+
+  private ServiceStatusEvaluationStage evaluatePendingOrStartingStatusCode(Plan evaluatePlan,
+                                                                          ServiceStatusCode pending,
```
Comment (kvish) [inline sdk/scheduler/src/main/java/com/mesosphere/sdk/http/endpoints/HealthResource.java]: formatting here

File: sdk/scheduler/src/main/java/com/mesosphere/sdk/http/endpoints/HealthResource.java [hunk 9]
```diff
@@ -1,79 +1,582 @@
...
-import java.util.Collection;
+import java.util.Optional;
+import java.util.Set;
 import java.util.stream.Collectors;
 
+// CHECKSTYLE:OFF LineLengthCheck
+
 /**
- * A read-only API for checking whether the service is healthy.
+ * This class returns a single code representing the status of the service
```
Comment (takirala) [inline sdk/scheduler/src/main/java/com/mesosphere/sdk/http/endpoints/HealthResource.java]: I am inclined to say to delete this javadoc, The `ServiceStatusCode` enum would serve as single source of truth.

File: sdk/scheduler/src/main/java/com/mesosphere/sdk/http/endpoints/HealthResource.java [hunk 10]
```diff
@@ -1,79 +1,582 @@
...
+ * 203, 2, Deploying:Waiting_User
+ * 206, 3, Degraded
+ * 203, 4, Recovering:Pending
+ * 205, 4, Recovering:Starting
+ * 320, 5, Backing Up
+ * 321, 5, Restoring
+ * 326, 6, Upgrade/Rollback/Downgrade
+ * 503,  , Service Unavailable (Priority Undefined)
  */
 @Path("/v1/health")
```
Comment (takirala) [inline sdk/scheduler/src/main/java/com/mesosphere/sdk/http/endpoints/HealthResource.java]: Have you tested this in clusters going back to 1.10? (Not sure if mesos treats the codes differently across DC/OS versions)

File: sdk/scheduler/src/test/java/com/mesosphere/sdk/http/endpoints/HealthResourceTest.java [hunk 11]
```diff
@@ -1,80 +1,1171 @@
...
 
-    @Test
-    public void testPlanComplete() {
-        when(mockPlan1.getErrors()).thenReturn(Collections.emptyList());
-        when(mockPlan1.isComplete()).thenReturn(true);
-        when(mockPlan2.getErrors()).thenReturn(Collections.emptyList());
-        when(mockPlan2.isComplete()).thenReturn(true);
-        checkHealthStatus(Response.Status.OK);
+  @Before
+  public void setupTest() throws Exception {
```
Comment (takirala) [inline sdk/scheduler/src/test/java/com/mesosphere/sdk/http/endpoints/HealthResourceTest.java]: wow mockito makes this `throws Exception` ?!

Comment (kaiwalyajoshi) [reply sdk/scheduler/src/test/java/com/mesosphere/sdk/http/endpoints/HealthResourceTest.java]: good catch! fixed

**Comment by takirala** [comment]
Also, If possible, add a note to the change log (that we started recently) on how the health endpoint changed.

File: sdk/scheduler/src/main/java/com/mesosphere/sdk/http/endpoints/HealthResource.java (hunk 10 above)
Comment (kaiwalyajoshi) [reply sdk/scheduler/src/main/java/com/mesosphere/sdk/http/endpoints/HealthResource.java]: Running tests on a 1.10 CCM cluster now to verify.

Comment (kaiwalyajoshi) [reply sdk/scheduler/src/main/java/com/mesosphere/sdk/http/endpoints/HealthResource.java]: Ran hello-world sanity on CCM 1.10.9 cluster, other than tests which involve setup via SSH keys (test_volumes, test_zzzrecovery_py), tests passed.

**Pull request merged**